lib/functions/OverloadTree.ts

```diff
@@ -170,24 +168,11 @@ export class OverloadTree {
       nextTree = newNode;
     }
     nextTree._addOverload(_argumentTypes, func, promotionCount);
-    // Defined by https://www.w3.org/TR/xpath-31/#promotion .
-    // e.g. When a function takes a string, it can also accept a XSD_ANY_URI if it's cast first.
-    // TODO: When promoting decimal type a cast needs to be preformed.
-    if (argumentType === TypeURL.XSD_STRING) {
-      this.addPromotedOverload(TypeURL.XSD_ANY_URI, func, arg =>
-        string(arg.str()), _argumentTypes, promotionCount);
-    }
-    // TODO: in case of decimal a round needs to happen.
-    if (argumentType === TypeURL.XSD_DOUBLE) {
-      this.addPromotedOverload(TypeURL.XSD_FLOAT, func, arg =>
-        double((<E.NumericLiteral>arg).typedValue), _argumentTypes, promotionCount);
-      this.addPromotedOverload(TypeURL.XSD_DECIMAL, func, arg =>
-        double((<E.NumericLiteral>arg).typedValue), _argumentTypes, promotionCount);
-    }
-    if (argumentType === TypeURL.XSD_FLOAT) {
-      this.addPromotedOverload(TypeURL.XSD_DECIMAL, func, arg =>
-        float((<E.NumericLiteral>arg).typedValue), _argumentTypes, promotionCount);
-    }
+
+    typePromotion[argumentType]?.forEach(ret =>
+      this.addPromotedOverload(
+        ret.typeToPromote, func, ret.conversionFunction, _argumentTypes, promotionCount,
+      ));
```

Comment:
Why was this change needed exactly?

Reply:
Ah yes. It was not. When starting this PR we talked about using some kind of type promotion. That's when I wrote this. The PR shifted to a whole new thing since then. I still think this new way is the better way. TypePromotion is a type handling problem. However, it does not being in this PR. Should I make a separate one for this?

Reply:
Type promotion is a finicky but important thing. A separate PR would be nice.

Reply:
I agree with going for a separate PR on this 👍